File: client-sdk/go/modules/rofl/rofl.go
package rofl

import (
	"context"
	"fmt"

	"github.com/oasisprotocol/oasis-core/go/common/cbor"

	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/client"
	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/types"
)

var (
	// Callable methods.
	methodCreate   = types.NewMethodName("rofl.Create", Create{})
	methodUpdate   = types.NewMethodName("rofl.Update", Update{})
	methodRemove   = types.NewMethodName("rofl.Remove", Remove{})
	methodRegister = types.NewMethodName("rofl.Register", Register{})

	// Queries.
	methodApp             = types.NewMethodName("rofl.App", AppQuery{})
	methodApps            = types.NewMethodName("rofl.Apps", nil)
	methodAppInstance     = types.NewMethodName("rofl.AppInstance", AppInstanceQuery{})
	methodAppInstances    = types.NewMethodName("rofl.AppInstances", AppQuery{})
	methodParameters      = types.NewMethodName("rofl.Parameters", nil)
	methodStakeThresholds = types.NewMethodName("rofl.StakeThresholds", nil)
)

// V1 is the v1 rofl module interface.
type V1 interface {
	client.EventDecoder

	// Create generates a rofl.Create transaction.
	Create(policy AppAuthPolicy) *client.TransactionBuilder

	// Update generates a rofl.Update transaction.
	Update(id AppID, policy AppAuthPolicy, admin *types.Address) *client.TransactionBuilder

	// Remove generates a rofl.Remove transaction.
	Remove(id AppID) *client.TransactionBuilder

	// App queries the given application configuration.
	App(ctx context.Context, round uint64, id AppID) (*AppConfig, error)

	// Apps queries all application configurations.
	Apps(ctx context.Context, round uint64) ([]*AppConfig, error)

	// AppInstance queries a specific registered instance of the given application.
	AppInstance(ctx context.Context, round uint64, id AppID, rak types.PublicKey) (*Registration, error)

	// AppInstances queries the registered instances of the given application.
	AppInstances(ctx context.Context, round uint64, id AppID) ([]*Registration, error)

	// StakeThresholds queries the stake information for managing ROFL.
	StakeThresholds(ctx context.Context, round uint64) (*StakeThresholds, error)

	// Parameters queries the module parameters.
	Parameters(ctx context.Context, round uint64) (*Parameters, error)

	// GetEvents returns all rofl events emitted in a given block.
	GetEvents(ctx context.Context, round uint64) ([]*Event, error)
}

type v1 struct {
	rc client.RuntimeClient
}

// Implements V1.
func (a *v1) Create(policy AppAuthPolicy) *client.TransactionBuilder {
	return client.NewTransactionBuilder(a.rc, methodCreate, &Create{
		Policy: policy,
	})
}

// Implements V1.
func (a *v1) Update(id AppID, policy AppAuthPolicy, admin *types.Address) *client.TransactionBuilder {
	return client.NewTransactionBuilder(a.rc, methodUpdate, &Update{
		ID:     id,
		Policy: policy,
		Admin:  admin,
	})
}

// Implements V1.
func (a *v1) Remove(id AppID) *client.TransactionBuilder {
	return client.NewTransactionBuilder(a.rc, methodRemove, &Remove{
		ID: id,
	})
}

// Implements V1.
func (a *v1) App(ctx context.Context, round uint64, id AppID) (*AppConfig, error) {
	var appCfg AppConfig
	err := a.rc.Query(ctx, round, methodApp, AppQuery{ID: id}, &appCfg)
	if err != nil {
		return nil, err
	}
	return &appCfg, nil
}

// Implements V1.
func (a *v1) Apps(ctx context.Context, round uint64) ([]*AppConfig, error) {
	var apps []*AppConfig
	err := a.rc.Query(ctx, round, methodApps, nil, &apps)
	if err != nil {
		return nil, err
	}
	return apps, nil
}

// Implements V1.
func (a *v1) AppInstance(ctx context.Context, round uint64, id AppID, rak types.PublicKey) (*Registration, error) {
	var instance Registration
	err := a.rc.Query(ctx, round, methodAppInstance, AppInstanceQuery{App: id, RAK: rak}, &instance)
	if err != nil {
		return nil, err
	}
	return &instance, nil
}

// Implements V1.
func (a *v1) AppInstances(ctx context.Context, round uint64, id AppID) ([]*Registration, error) {
	var instances []*Registration
	err := a.rc.Query(ctx, round, methodAppInstances, AppQuery{ID: id}, &instances)
	if err != nil {
		return nil, err
	}
	return instances, nil
}

// Implements V1.
func (a *v1) StakeThresholds(ctx context.Context, round uint64) (*StakeThresholds, error) {
	var thresholds StakeThresholds
	err := a.rc.Query(ctx, round, methodStakeThresholds, nil, &thresholds)
	if err != nil {
		return nil, err
	}
	return &thresholds, nil
}

// Implements V1.
func (a *v1) Parameters(ctx context.Context, round uint64) (*Parameters, error) {
	var params Parameters
	err := a.rc.Query(ctx, round, methodParameters, nil, &params)
	if err != nil {
		return nil, err
	}
	return &params, nil
}

// Implements V1.
func (a *v1) GetEvents(ctx context.Context, round uint64) ([]*Event, error) {
	rawEvs, err := a.rc.GetEventsRaw(ctx, round)
	if err != nil {
		return nil, err
	}

	evs := make([]*Event, 0)
	for _, rawEv := range rawEvs {
		ev, err := a.DecodeEvent(rawEv)
		if err != nil {
			return nil, err
		}
		for _, e := range ev {
			evs = append(evs, e.(*Event))
		}
	}

	return evs, nil
}

// Implements client.EventDecoder.
func (a *v1) DecodeEvent(event *types.Event) ([]client.DecodedEvent, error) {
	return DecodeEvent(event)
}

// DecodeEvent decodes a rofl event.
func DecodeEvent(event *types.Event) ([]client.DecodedEvent, error) {
	if event.Module != ModuleName {
		return nil, nil
	}
	var events []client.DecodedEvent
	switch event.Code {
	case AppCreatedEventCode:
		var evs []*AppCreatedEvent
		if err := cbor.Unmarshal(event.Value, &evs); err != nil {
			return nil, fmt.Errorf("decode rofl app created event value: %w", err)
		}
		for _, ev := range evs {
			events = append(events, &Event{AppCreated: ev})
		}
	case AppUpdatedEventCode:
		var evs []*AppUpdatedEvent
		if err := cbor.Unmarshal(event.Value, &evs); err != nil {
			return nil, fmt.Errorf("decode rofl app updated event value: %w", err)
		}
		for _, ev := range evs {
			events = append(events, &Event{AppUpdated: ev})
		}
	case AppRemovedEventCode:
		var evs []*AppRemovedEvent
		if err := cbor.Unmarshal(event.Value, &evs); err != nil {
			return nil, fmt.Errorf("decode rofl app removed event value: %w", err)
		}
		for _, ev := range evs {
			events = append(events, &Event{AppRemoved: ev})
		}
	case InstanceRegisteredEventCode:
		var evs []*InstanceRegisteredEvent
		if err := cbor.Unmarshal(event.Value, &evs); err != nil {
			return nil, fmt.Errorf("decode rofl instance registered event value: %w", err)
		}
		for _, ev := range evs {
			events = append(events, &Event{InstanceRegistered: ev})
		}
	default:
		return nil, fmt.Errorf("invalid rofl event code: %v", event.Code)
	}
	return events, nil
}

// NewV1 generates a V1 client helper for the rofl module.
func NewV1(rc client.RuntimeClient) V1 {
	return &v1{rc: rc}
}

// NewCreateTx generates a new rofl.Create transaction.
func NewCreateTx(fee *types.Fee, body *Create) *types.Transaction {
	return types.NewTransaction(fee, methodCreate, body)
}

// NewUpdateTx generates a new rofl.Update transaction.
func NewUpdateTx(fee *types.Fee, body *Update) *types.Transaction {
	return types.NewTransaction(fee, methodUpdate, body)
}

// NewRemoveTx generates a new rofl.Remove transaction.
func NewRemoveTx(fee *types.Fee, body *Remove) *types.Transaction {
	return types.NewTransaction(fee, methodRemove, body)
}

// NewRegisterTx generates a new rofl.Register transaction.
func NewRegisterTx(fee *types.Fee, body *Register) *types.Transaction {
	return types.NewTransaction(fee, methodRegister, body)
}

File: client-sdk/ts-web/rt/src/rofl.ts
import * as oasis from '@oasisprotocol/client';
import * as event from './event';
import * as transaction from './transaction';
import * as types from './types';
import * as wrapper from './wrapper';
import * as nacl from 'tweetnacl';
import * as mraeDeoxysii from './mrae/deoxysii';
import * as deoxysii from '@oasisprotocol/deoxysii';

/**
 * Unique module name.
 */
export const MODULE_NAME = 'rofl';

// Callable methods.
export const METHOD_CREATE = 'rofl.Create';
export const METHOD_UPDATE = 'rofl.Update';
export const METHOD_REMOVE = 'rofl.Remove';
export const METHOD_REGISTER = 'rofl.Register';

// Queries.
export const METHOD_APP = 'rofl.App';
export const METHOD_APPS = 'rofl.Apps';
export const METHOD_APP_INSTANCE = 'rofl.AppInstance';
export const METHOD_APP_INSTANCES = 'rofl.AppInstances';
export const METHOD_PARAMETERS = 'rofl.Parameters';
export const METHOD_STAKE_THRESHOLDS = 'rofl.StakeThresholds';

// Events.
export const EVENT_APP_CREATED_CODE = 1;
export const EVENT_APP_UPDATED_CODE = 2;
export const EVENT_APP_REMOVED_CODE = 3;
export const EVENT_INSTANCE_REGISTERED_CODE = 4;

export const ADDRESS_PREFIX = 'rofl';
export function toBech32(appID: types.AppID) {
    if (appID.length !== 21) {
        throw new Error(`wrong address length: ${appID}`);
    }
    return oasis.address.toBech32(ADDRESS_PREFIX, appID);
}
export function fromBech32(str: string): types.AppID {
    if (str.length !== 41 + ADDRESS_PREFIX.length) {
        // prevent e.g. rofl1qqqqqqqtwy44z
        throw new Error(`wrong address length: ${str}`);
    }
    return oasis.address.fromBech32(ADDRESS_PREFIX, str);
}

export class Wrapper extends wrapper.Base {
    constructor(runtimeID: Uint8Array) {
        super(runtimeID);
    }

    /** Create a new ROFL application. */
    callCreate() {
        return this.call<types.RoflCreate, void>(METHOD_CREATE);
    }

    /** Update a ROFL application. */
    callUpdate() {
        return this.call<types.RoflUpdate, void>(METHOD_UPDATE);
    }

    /** Remove a ROFL application. */
    callRemove() {
        return this.call<types.RoflRemove, void>(METHOD_REMOVE);
    }

    /** Register a new ROFL App instance replica. */
    callRegister() {
        return this.call<types.RoflRegister, void>(METHOD_REGISTER);
    }

    /** Returns the configuration for the given ROFL application. */
    queryApp() {
        return this.query<types.RoflAppQuery, types.RoflAppConfig>(METHOD_APP);
    }

    /** Returns all ROFL app configurations. */
    queryApps() {
        return this.query<void, types.RoflAppConfig[]>(METHOD_APPS);
    }

    /** Returns a specific registered instance replica for the given ROFL application. */
    queryAppInstance() {
        return this.query<types.RoflAppInstanceQuery, types.RoflRegistration>(METHOD_APP_INSTANCE);
    }

    /** Returns a list of all registered instances replicas for the given ROFL application. */
    queryAppInstances() {
        return this.query<types.RoflAppQuery, types.RoflRegistration[]>(METHOD_APP_INSTANCES);
    }

    /** Returns the minimum stake thresholds for managing ROFL. */
    queryStakeThresholds() {
        return this.query<void, types.RoflStakeThresholds>(METHOD_STAKE_THRESHOLDS);
    }

    /** Queries the module parameters */
    queryParameters() {
        return this.query<void, void>(METHOD_PARAMETERS);
    }
}

export function moduleEventHandler(codes: {
    [EVENT_APP_CREATED_CODE]?: event.Handler<types.RoflAppCreatedEvent>;
    [EVENT_APP_UPDATED_CODE]?: event.Handler<types.RoflAppUpdatedEvent>;
    [EVENT_APP_REMOVED_CODE]?: event.Handler<types.RoflAppRemovedEvent>;
    [EVENT_INSTANCE_REGISTERED_CODE]?: event.Handler<types.RoflInstanceRegisteredEvent>;
}) {
    return [MODULE_NAME, codes] as event.ModuleHandler;
}

export type TransactionCallHandlers = {
    [METHOD_CREATE]?: transaction.CallHandler<types.RoflCreate>;
    [METHOD_UPDATE]?: transaction.CallHandler<types.RoflUpdate>;
    [METHOD_REMOVE]?: transaction.CallHandler<types.RoflRemove>;
    [METHOD_REGISTER]?: transaction.CallHandler<types.RoflRegister>;
};

/** SecretEnvelope is the envelope used for storing encrypted secrets. */
export interface SecretEnvelope {
    /** Pk is the ephemeral public key used for X25519. */
    pk: Uint8Array;
    /** Nonce. */
    nonce: Uint8Array;
    /** Name is the encrypted secret name. */
    name: Uint8Array;
    /** Value is the encrypted secret value. */
    value: Uint8Array;
}

/**
 * EncryptSecret encrypts the given secret given its plain-text name and value together with the
 * secrets encryption key (SEK) obtained for the given application. Returns the Base64-encoded
 * value that can be used in the configuration.
 *
 * Based on https://github.com/oasisprotocol/cli/blob/1fd79c1/build/rofl/secrets.go#L57
 * and https://github.com/oasisprotocol/oasis-sdk/blob/3177bbc/client-sdk/ts-web/rt/src/callformat.ts#L52
 */
export function encryptSecret(
    name: string,
    value: Uint8Array,
    app_sek: types.RoflAppConfig['sek'],
): string {
    // Generate ephemeral X25519 key pair.
    const keyPair = nacl.box.keyPair();
    const pk = keyPair.publicKey;
    const sk = keyPair.secretKey;

    // Generate random nonce.
    const nonce = new Uint8Array(deoxysii.NonceSize);
    crypto.getRandomValues(nonce);

    // Seal plain-text secret name and value.
    const encName = mraeDeoxysii.boxSeal(
        nonce,
        oasis.misc.fromString(name),
        oasis.misc.fromString('name'),
        app_sek,
        sk,
    );
    const encValue = mraeDeoxysii.boxSeal(
        nonce,
        value,
        oasis.misc.fromString('value'),
        app_sek,
        sk,
    );

    const envelope: SecretEnvelope = {
        pk: pk,
        nonce: nonce,
        name: encName,
        value: encValue,
    };

    const data = oasis.misc.toCBOR(envelope);
    return oasis.misc.toBase64(data);
}
